lec08-stability.Rmd

```diff
@@ -1,4 +1,589 @@
+you calculate the rates of change using the variables from the previous time step. For 
+example, in the predator-prey model, both $\Delta x$ and $\Delta y$ depend on $x$ and $y$, 
+and it's important that you calculate $\Delta x$ and $\Delta y$ *before* updating the 
+values of $x$ and $y$.
+
+Here's an example. 
+
+```{r}
+# Define predator-prey equations
+
+dX_dt <- function(X, Y, a, b) {
+  # X: number of prey
+  # Y: number of predators
+  # a: growth rate of prey
+  # b: rate at which predators eat prey
+  
+  return(a * X - b * Y * X)
+}
+
+dY_dt <- function(X, Y, c, d) {
+  # X: number of prey
+  # Y: number of predators
+  # c: growth rate of predators from eating prey
+  # d: death rate for predators
+  
+  return (c * X - d * Y)
+}
+```
+
+```{r}
+# parameters
+a <- 5
+b <- 1
+c <- 1
+d <- 0.2
+
+state <- c(X = 2, Y = 2)
+
+dt <- 0.01 # timestep - the smaller, the better
+tmax <- 30 # the total time we want to numerically solve for
+points <- tmax/dt # the number of data points in the simulation - add 1 so that we can start at t=0
+
+# vectors to store values of X, Y, and t at each timestep:
+X_vector <- numeric(points) # prey population size
+Y_vector <- numeric(points) # predator population size
+t_vector <- seq(0, tmax - dt, by = dt) # time vector
+
+# initial condition
+X0 <- 2
+Y0 <- 2
+X_vector[1] <- X0
+Y_vector[1] <- Y0
+
+X <- X0 # initialize variable X
+Y <- Y0 # initialize variable Y
+```
+
+This is the right way to do it.
+
+```{r}
+for (i in 2:points) {
+  # start at 2 because the initial state is at position 1
+  
+  # first calculate BOTH dX and dY
+  dX <- dX_dt(X = X, Y = Y, a = a, b = b) * dt
+  dY <- dY_dt(X = X, Y = Y, c = c, d = d) * dt
+  
+  # now update BOTH X and Y
+  X <- X + dX 
+  Y <- Y + dY
+  X_vector[i] <- X
+  Y_vector[i] <- Y
+}
+```
+
+This is the wrong way.
+
+```{r}
+for (i in 2:points) {
+  # start at 2 because the initial state is at position 1
+  
+  # calculate dX and then update X (wrong)
+  dX <- dX_dt(X = X, Y = Y, a = a, b = b) * dt
+  X < X + dX
+  
+  # calculate dY and then update Y - this will use the new value of X instead of the old one.
+  dY <- dY_dt(X = X, Y = Y, c = c, d = d) * dt
+  Y <- Y + dY
+  
+  X_vector[i] <- X
+  Y_vector[i] <- Y
+}
+```
+
+## Phase portraits in two dimensions
+
+#### Challenge
+
+What were the axes of the one-dimensional phase portraits we drew? What would the axes
+of a phase portrait for the predator-prey system above be?
+
+```{r, eval=FALSE, echo=FALSE}
+# Challenge solution
+
+# The axis was N in the 1D version. 
+# For a 2D system with variables x and y, the two axes would be x and y.
+```
+
+Continuing with the predator-prey example: instead of plotting our numerical trajectory as $X$ and 
+$Y$ vs. $t$, we'll plot $Y$ vs. $X$: predator population size vs. prey population size. 
+
+```{r}
+ggplot(result) +
+  geom_point(aes(x = X, y = Y)) # geom_line connects the dots strangely
```

Comment:
Is it supposed to look strange?

Reply:
Uh no?

Reply:
Ah, right. Line assumes a continuous link for lines for each increase in x. What you want is `geom_path` which links based on the x and y pairing.

Reply:
You could always change the gif to use path as well.

Reply:
ooooohhhhhh
!!!!

Reply:
She uses `geom_point` not `geom_line`, so it should be fine? It doesn't look strange to me (I think), see below.

![image](https://user-images.githubusercontent.com/4560057/31104828-72e945c0-a7ad-11e7-9593-f31698d1d578.png)

You could color by time to clarify that there another variable changing:

```r
ggplot(result) +
  geom_point(aes(x = X, y = Y, color = time)) # geom_line connects the dots strangely
```

![image](https://user-images.githubusercontent.com/4560057/31104932-411be376-a7ae-11e7-83ce-2bbd022586c6.png)

Reply:
Oh just understood that your comment about `geom_line` was referring to the comment in the code. I like the dots though!

Reply:
I LOVE the colouring by time!!!!